presto-main/src/main/java/com/facebook/presto/operator/aggregation/AbstractMinMaxNAggregation.java

```diff
@@ -0,0 +1,175 @@
+    private final Signature signature;
+
+    protected AbstractMinMaxNAggregation(String name, Function<Type, BlockComparator> typeToComparator)
+    {
+        checkNotNull(name);
+        checkNotNull(typeToComparator);
+        this.name = name;
+        this.typeToComparator = typeToComparator;
+        this.signature = new Signature(name, ImmutableList.of(orderableTypeParameter("E")), "array<E>", ImmutableList.of("E", StandardTypes.BIGINT), false, false);
+    }
+
+    @Override
+    public Signature getSignature()
+    {
+        return signature;
+    }
+
+    @Override
+    public FunctionInfo specialize(Map<String, Type> types, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
+    {
+        Type type = types.get("E");
+        Signature signature = new Signature(name, TypeUtils.parameterizedTypeName("array", type.getTypeSignature()), type.getTypeSignature(), BIGINT.getTypeSignature());
+        InternalAggregationFunction aggregation = generateAggregation(type);
+        return new FunctionInfo(signature, getDescription(), aggregation);
+    }
+
+    protected InternalAggregationFunction generateAggregation(Type type)
+    {
+        DynamicClassLoader classLoader = new DynamicClassLoader(AbstractMinMaxNAggregation.class.getClassLoader());
+
+        BlockComparator comparator = typeToComparator.apply(type);
+        List<Type> inputTypes = ImmutableList.of(type, BIGINT);
+        MinMaxNStateSerializer stateSerializer = new MinMaxNStateSerializer(comparator, type);
+        Type intermediateType = stateSerializer.getSerializedType();
+        ArrayType finalType = new ArrayType(type);
+
+        List<ParameterMetadata> inputParameterMetadata =  ImmutableList.of(
+                new ParameterMetadata(STATE),
+                new ParameterMetadata(BLOCK_INPUT_CHANNEL, type),
+                new ParameterMetadata(INPUT_CHANNEL, BIGINT),
+                new ParameterMetadata(BLOCK_INDEX));
+
+        AggregationMetadata metadata = new AggregationMetadata(
+                generateAggregationName(name, type, inputTypes),
+                inputParameterMetadata,
+                INPUT_FUNCTION.bindTo(comparator).bindTo(type),
+                null,
+                null,
+                COMBINE_FUNCTION,
+                OUTPUT_FUNCTION.bindTo(finalType),
+                MinMaxNState.class,
+                stateSerializer,
+                new MinMaxNStateFactory(),
+                finalType,
+                false);
+
+        GenericAccumulatorFactoryBinder factory = new AccumulatorCompiler().generateAccumulatorFactoryBinder(metadata, classLoader);
+        return new InternalAggregationFunction(name, inputTypes, intermediateType, finalType, true, false, factory);
+    }
+
+    public static void input(BlockComparator comparator, Type type, MinMaxNState state, Block block, long n, int blockIndex)
+    {
+        if (block.isNull(blockIndex)) {
```

Comment:
this can never happen, since you marked this as a `BLOCK_INPUT_CHANNEL` it will skip null rows